### DISCUSSION TITLE
Vulkan backend & ggml-vulkan-shaders.hpp

### DISCUSSION THREAD

**Discussion by mjgarton** [ggml-org/llama.cpp#7323] [Q&A] [answered]
I am just wondering how the contents of ggml-vulkan-shaders.hpp was generated?  I would like to learn about how the vulkan backend works.

If I understand correctly, this is compiled spirv shader code.  Why store the compiled code as bytes here, rather than storing source and generating the binary at build time?  Would this improve maintainability & transparency?

PS. I am fairly new to llama.cpp, GPUs, vulkan, shaders and ML in general, so apologies if my questions are somewhat confused.

**Comment by teleprint-me** [comment, marked as answer]
The shaders are generated using the `ggml_vk_generate_shaders.py` script. You shouldn't need to use it unless the source is modified. The shaders are included to smooth out compilation and runtime.

You can see the rationale behind most of the design decisions in the initial PR #2059.

**Comment by uap-coder** [comment]
can you make an action that will generate this from the current Vulkan files, I've been trying. it doesn't work without a computer older than 2015. I cant find or make this any other way ggml-vulkan-shaders.hpp
